upf/upf_pfcp_api.c

```diff
@@ -1542,9 +1547,34 @@ handle_create_far (upf_session_t * sx, pfcp_create_far_t * create_far,
 
 	    ip_udp_gtpu_rewrite (&create->forward, fib_index, is_ip4);
 	  }
-	//TODO: transport_level_marking
-	//TODO: forwarding_policy
-	//TODO: header_enrichment
+// TODO: transport_level_marking
+// forwarding_policy >> oln: Implementation
+
+	if (ISSET_BIT (far->forwarding_parameters.grp.fields,
+		       FORWARDING_PARAMETERS_FORWARDING_POLICY))
+	  {
+	    policy_id =
+	      far->forwarding_parameters.forwarding_policy.identifier;
+	    hash_ptr = hash_get_mem (gtm->forwarding_policy_by_id, policy_id);
+	    if (hash_ptr)	// validate if policy is preconfigured
+	      {
+		create->forward.flags |= FAR_F_FORWARDING_POLICY;
+		create->forward.forwarding_policy.identifier =
+		  far->forwarding_parameters.forwarding_policy.identifier;
+		fp_entry =
+		  pool_elt_at_index (gtm->upf_forwarding_policies,
+				     hash_ptr[0]);
+		create->forward.fp_pool_index = hash_ptr[0];
+		fp_entry->ref_cnt++;
+	      }
+	    else
+	      {
+		upf_debug
+		  ("###### Forwarding policy id %v is not preconfigured at UPF ######",
+		   far->forwarding_parameters.forwarding_policy.identifier);
+		create->forward.fp_pool_index = ~0;
```

Comment:
this should instead return a "Invalid Forwarding Policy" error cause.

Reply:
Done